### PR TITLE
Add IotCentral ARM SDK

### PR DESCRIPTION
**Pull request by emgarten**
## Description

* Add C# SDK for Iot Central for the first time
* Add Iot Central SDK tests

API Spec PR: https://github.com/Azure/azure-rest-api-specs/pull/3185

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. *(@emgarten: Uses latest, let me know if this should be changed. I based the generate files off other SDKs in the repo)*
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. *(@emgarten: this is set in the csproj)*

### CODE REVIEW

File: src/SDKs/IotCentral/Management.IotCentral/generate.ps1 [hunk 1]
```diff
@@ -0,0 +1 @@
+powershell.exe -ExecutionPolicy Bypass -NoLogo -NonInteractive -NoProfile -File "$(split-path $SCRIPT:MyInvocation.MyCommand.Path -parent)\..\..\..\..\tools\generateTool.ps1" -ResourceProvider "iotcentral/resource-manager" -PowershellInvoker  -AutoRestVersion "latest"
```
Comment (dsgouda) [inline src/SDKs/IotCentral/Management.IotCentral/generate.ps1]: generate.ps1 uses powershell cmdlet now, please take a look at [this](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/Compute/Management.Compute/generate.ps1) and make appropriate changes and regenerate the code.

File: src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		<Description>Provides management capabilities for Microsoft Azure IotCentral.</Description>
+		<AssemblyTitle>Microsoft Azure IotCentral Management</AssemblyTitle>
+		<AssemblyName>Microsoft.Azure.Management.IotCentral</AssemblyName>
+		<Version>1.0.0-preview</Version>
+		<PackageTags>Microsoft Azure IotCentral;IotCentral management;IotCentral;</PackageTags>
+		<PackageReleaseNotes>
+			<![CDATA[
+        This is a public preview release of the Azure IoT Central SDK.
+      ]]>
+		</PackageReleaseNotes>
```
Comment (dsgouda) [inline src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj]: nit: indentation is usually 2 spaces.

File: src/SDKs/IotCentral/Management.IotCentral/generate.cmd [hunk 3]
```diff
@@ -0,0 +1,7 @@
+::
+:: Microsoft Azure SDK for Net - Generate library code
+:: Copyright (C) Microsoft Corporation. All Rights Reserved.
+::
+
+@echo off
+call %~dp0..\..\..\..\tools\generate.cmd iotcentral/resource-manager %*
```
Comment (dsgouda) [inline src/SDKs/IotCentral/Management.IotCentral/generate.cmd]: generate.cmds are not needed anymore, all code generation must go through generate.ps1s

**Review by dsgouda — CHANGES_REQUESTED** [review]
LGTM for the most part, left a few comments.

File: src/SDKs/IotCentral/Management.IotCentral/generate.ps1 (hunk 1 above)
Comment (emgarten) [reply src/SDKs/IotCentral/Management.IotCentral/generate.ps1]: I've updated the ps1 script and generated the files again to make sure everything is in sync.

File: src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj (hunk 2 above)
Comment (emgarten) [reply src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj]: Fixed, there was a mix of tabs and spaces here.

File: src/SDKs/IotCentral/Management.IotCentral/generate.cmd (hunk 3 above)
Comment (emgarten) [reply src/SDKs/IotCentral/Management.IotCentral/generate.cmd]: removed

**Comment by emgarten** [comment]
Thanks for the feedback @dsgouda I've updated the PR

**Comment by emgarten** [comment]
//cc @vishwam

File: src/SDKs/IotCentral/Management.IotCentral/generate.ps1 [hunk 4]
```diff
@@ -0,0 +1 @@
+Start-AutoRestCodeGeneration -ResourceProvider "iotcentral/resource-manage" -AutoRestVersion "latest" -SdkRootDirectory $PSScriptRoot
```
Comment (dsgouda) [inline src/SDKs/IotCentral/Management.IotCentral/generate.ps1]: Please fix the `ResourceProvider` here, should be `iotcentral/resource-manager`

File: src/SDKs/_metadata/iotcentral_resource-manage.txt [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿Installing AutoRest version: latest
+
```
Comment (dsgouda) [inline src/SDKs/_metadata/iotcentral_resource-manage.txt]: Please delete this file and regenerate code

**Review by dsgouda — CHANGES_REQUESTED** [review]
Typo resulted in incorrect log file being generated.

File: src/SDKs/_metadata/iotcentral_resource-manage.txt (hunk 5 above)
Comment (emgarten) [reply src/SDKs/_metadata/iotcentral_resource-manage.txt]: I've cleaned this up

File: src/SDKs/IotCentral/Management.IotCentral/generate.ps1 (hunk 4 above)
Comment (emgarten) [reply src/SDKs/IotCentral/Management.IotCentral/generate.ps1]: Good catch, I've updated this and regenerated everything

File: src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  <!-- Please do not move/edit code below this line -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('AzSdk.reference.props'))" />
+  <!-- Please do not move/edit code below this line -->
+
+  <PropertyGroup>
+    <PackageId>Microsoft.Azure.Management.IotCentral</PackageId>
+    <Description>Provides management capabilities for Microsoft Azure IotCentral.</Description>
+    <AssemblyTitle>Microsoft Azure IotCentral Management</AssemblyTitle>
+    <AssemblyName>Microsoft.Azure.Management.IotCentral</AssemblyName>
+    <Version>1.0.0-preview</Version>
```
Comment (dsgouda) [inline src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj]: If you are planning to release a preview version, the versioning must start from `0.9.0-preview`

Comment (dsgouda) [reply src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj]: I missed this earlier but we also need `Assembly.cs` files similar to this for both the projects with the assembly versions properly set.

**Review by dsgouda — CHANGES_REQUESTED** [review]
Almost there, please address the remaining comments.

File: src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj (hunk 6 above)
Comment (emgarten) [reply src/SDKs/IotCentral/Management.IotCentral/Microsoft.Azure.Management.IotCentral.csproj]: I've added Assembly.cs and changed the csproj version to `0.9.0-preview`

image

**Review by dsgouda — APPROVED** [review]
LGTM

**Comment by emgarten** [comment]
@dsgouda when will the package for this be on nuget.org?

**Comment by dsgouda** [comment]
@emgarten we don't publish the packages. Please read the instructions to publish the package [here](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/Documentation/code-generation-instructions.md).

**Pull request merged**